### PR TITLE
Fix Noise design matrix index in the fitter

### PR DESCRIPTION
**Pull request by luojing1211**
The fitter gets the noise design matrix from the stack design matrix using the indices. The stack design matrix's index is labeled as 
+++++++++++++++++++++++++++++++++++++++++
           |    offset | fit parameters | Noise design matrix
+++++++++++++++++++++++++++++++++++++++++
TOAs |

In the red noise residual calculation, the offset column is not taken into consideration. The start index of the noise design matrix is 1 smaller than the real start position. This PR fixes it.

### CODE REVIEW

**Comment by paulray** [comment]
This is one of the many reasons why the offset parameter should be *explicit* in the model and not implicit.

**Comment by luojing1211** [comment]
I agree. But we have not decided on how to handle the absolute phase fitting.

File: tests/test_gls_fitter.py [hunk 1]
```diff
@@ -60,6 +60,20 @@ def test_gls_fitter(self):
                     assert np.abs(v - val[0]) <= e, msg
                     assert np.abs(1 - val[1] / e) < 0.1, msg
 
+    def test_noise_design_matrix_index(self):
+        self.fit(full_cov=False)
```
Comment (aarchiba) [inline tests/test_gls_fitter.py:64]: Why is it necessary to do a fit here? That will be slow.

File: tests/test_gls_fitter.py [hunk 2]
```diff
@@ -60,6 +60,20 @@ def test_gls_fitter(self):
...
+        ntmpar = len(self.f.model.free_params)
+        Mn = self.f.model.noise_model_designmatrix(self.f.toas)
+        M, params, units = self.f.get_designmatrix()
+        M_stack = np.hstack((M, Mn))
+        noise_dims = self.f.model.noise_model_dimensions(self.f.toas)
+        # This is how the gls fitter get the noise design matrix. I don't know
+        # if there is some other way to test the index.
```
Comment (aarchiba) [inline tests/test_gls_fitter.py:71]: For noise model basis functions that are easy to describe - like the constant offset one, or ECORR for a simple data set (merge two fake data sets so the ECORR basis functions are pairs) - you can confirm that the basis functions are what you think they should be. I wrote some examples of this in the tests for the downhill fitters, if I recall correctly.

File: tests/test_gls_fitter.py [hunk 3]
```diff
@@ -60,6 +60,20 @@ def test_gls_fitter(self):
...
+        for comp in noise_dims.keys():
+            p0 = noise_dims[comp][0] + ntmpar + 1
+            p1 = p0 + noise_dims[comp][1]
+            assert np.all(M_stack[:, p0:p1] == Mn[:, p0 - ntmpar - 1 : p1 - ntmpar - 1])
```
Comment (aarchiba) [inline tests/test_gls_fitter.py:75]: I don't understand what relationship this is supposed to be testing.

File: tests/test_model_manual.py [hunk 4]
```diff
@@ -89,7 +89,7 @@ def test_valid_model(tmp_dir, func, name, expectation):
 def test_compare_get_model_new_and_old():
     m_new = get_model_new(parfile)
     m_old = get_model(parfile)
-
+    print(parfile)
```
Comment (aarchiba) [inline tests/test_model_manual.py:92]: Do not print. You can always run `pytest --pdb` and inspect local variables in a failing test.

File: tests/test_gls_fitter.py [hunk 5]
```diff
@@ -60,6 +60,20 @@ def test_gls_fitter(self):
                     assert np.abs(v - val[0]) <= e, msg
                     assert np.abs(1 - val[1] / e) < 0.1, msg
 
+    def test_noise_design_matrix_index(self):
```
Comment (aarchiba) [inline tests/test_gls_fitter.py:63]: Does this test fail when run against the code before the bug fix? I can't really tell what it is testing or whether it actually verifies correctness.

Comment (aarchiba) [inline tests/test_gls_fitter.py:63]: Did you also test the downhill GLS and the wideband (downhill and not) fitters for the same bug?

File: src/pint/fitter.py [hunk 6]
```diff
@@ -1432,6 +1432,8 @@ def fit_toas(self, maxiter=10, threshold=0, full_cov=False, **kwargs):
             noise_dims = self.model.noise_model_dimensions(self.toas)
             noise_resids = {}
             for comp in noise_dims.keys():
+                # The first column of designmatrix is "offset", add 1 to match
+                # the indices of noise designmatrix
                 p0 = noise_dims[comp][0] + ntmpar
```
Comment (aarchiba) [inline src/pint/fitter.py:1437]: Should this line see the same change as the others?

File: src/pint/fitter.py [hunk 7]
```diff
@@ -1432,6 +1432,8 @@ def fit_toas(self, maxiter=10, threshold=0, full_cov=False, **kwargs):
             noise_dims = self.model.noise_model_dimensions(self.toas)
             noise_resids = {}
             for comp in noise_dims.keys():
```
Comment (aarchiba) [inline src/pint/fitter.py:1434]: Just use `for comp in noise_dims:` if you're changing the code anyway.

**Comment by luojing1211** [comment]
Our current test does not complain about the mismatch of indices. The only place of the usage is computing the red noise residual, shifting one index does not change the result dramatically. The test I put there is to use the same code from the fitter to produce the requested indices which are used to call the full design matrix. Then I compare if the sliced matrix the same as the noise design matrix.

File: tests/test_gls_fitter.py (hunk 2 above)
Comment (luojing1211) [reply tests/test_gls_fitter.py:71]: The problem is to get the noise model design matrix values from the full design matrix (Timingmodel design matrix + noise model design matrix). The noise dim calculation is ok. The problem is at the fitter. I don't have a good way to get P0, and P1, listed below from the fitter.

**Comment by aarchiba** [comment]
It is fairly easy to check red noise residuals if the basis functions are the top-hats you get from ECORR. This would explain some of the problems we've been having with whitened residuals. I wrote some relevant tests in #1000

File: tests/test_model_manual.py (hunk 4 above)
Comment (scottransom) [reply tests/test_model_manual.py:92]: I don't understand the "Do not print."  Why is that?  It gets suppressed during normal usage, but you can see if if you want by using `pytest -s`.  It seems (and has been) useful to me.

**Pull request merged**